File: stack.cs
class Stack
    {
        #region member variables
        private int top;
        private int[] data;
        #endregion

        public Stack(int maxSize)
        {
            top = -1;
            data = new int[maxSize];
        }

        public void Push(int num)
        {
            if (IsFull())
                throw new FullStackException();

            top += 1;
            data[top] = num;
        }

        public int Pop()
        {
            if (IsEmpty())
                throw new EmptyStackException();

            var topElem = data[top];
            top -= 1;

            return topElem;
        }

        public bool IsFull()
        {
            return top == data.Length-1;
        }

        public bool IsEmpty()
        {
            return top == -1;
        }
    }

    class FullStackException : Exception
    {
        public FullStackException() { }
    }

    class EmptyStackException : Exception
    {
        public EmptyStackException() { }
    }

    class StackTest
    {
        static void Main(string[] args)
        {
            var stack = new Stack(4);
            stack.Push(3);
            stack.Push(6);
            stack.Push(8);
            stack.Pop();
            stack.Push(5);
            stack.Pop();
            stack.Pop();
            stack.Pop();
            //stack.Pop();
            stack.Push(2);
            stack.Push(6);
            stack.Push(5);
            stack.Push(7);
            stack.Push(9);
        }
    }

File: stack.py
class Stack:
	def __init__(self, max_size):
		self.top = -1
		self.max_size = max_size
		self.stack = [None for _ in range(max_size)]
	
	def is_full(self):
		return self.top == self.max_size
	
	def is_empty(self):
		return self.top == 0

	def push(self, val):
		self.top += 1
		if self.is_full() == True:
			raise FullExeption
		self.stack[self.top] = val

	def pop(self):
		if self.is_empty() == True:
			raise EmptyExeption    		
		res = self.stack[self.top]
		self.top -=1
	
		return res

class FullExeption(Exception):
    def __init__(self):
        super().__init__('Stack is full')

class EmptyExeption(Exception):
    	def __init__(self):
    		super().__init__('Stack is empty')

##########		
# Adaptor Design pattern: 이미 만들어진 것(pop)을 다시 재사용
class Stack2:
	def __init__(self):
		self.stack = []
	
	def push(self, num):
		self.stack.append(num)

	def pop(self):
		if self.isEmpty():
			raise Exception
		else:
			return self.stack.pop()

	def isEmpty(self):
		return len(self.stack) == 0
